packages/flutter/lib/src/widgets/animated_scroll_view.dart

```diff
@@ -211,6 +211,341 @@ class AnimatedListState extends _AnimatedScrollViewState<AnimatedList> {
+  ///
+  /// This method does not create a dependency, and so will not cause rebuilding
+  /// when the state changes.
+  ///
+  /// See also:
+  ///
+  ///  * [of], a similar function that will throw if no [AnimatedListSeparated] ancestor
+  ///    is found.
+  static AnimatedListSeparatedState? maybeOf(BuildContext context) {
+    return context.findAncestorStateOfType<AnimatedListSeparatedState>();
+  }
+
+  @override
+  AnimatedListSeparatedState createState() => AnimatedListSeparatedState();
+
+  // Helper method to compute the actual child count when taking separators into account.
+  static int _computeChildCountWithSeparators(int itemCount) {
+    if (itemCount == 0) {
+      return 0;
+    }
+    return itemCount * 2 - 1;
+  }
+}
+
+/// The [AnimatedListSeparatedState] for [AnimatedListSeparated],
+/// a scrolling list container that animates items with separators when they are
+/// inserted or removed.
+///
+/// When an item is inserted with [insertItem] an animation begins running.
+/// The animation is passed to [AnimatedListSeparated.itemBuilder]
+/// and [AnimatedListSeparated.separatorBuilder] whenever the item's
+/// and separator's widgets are needed.
+///
+/// When multiple items are inserted with [insertAllItems] an animation begins running.
+/// The animation is passed to [AnimatedListSeparated.itemBuilder]
+/// and [AnimatedListSeparated.separatorBuilder] whenever
+/// the item's and separator's widgets are needed.
+///
+/// When an item is removed with [removeSeparatedItem] its animation
+/// as well as that of its corresponding separator are reversed.
+/// The removed item's animation is passed to
+/// the [removeSeparatedItem] itemBuilder parameter.
+/// The corresponding separator's animation is passed
+/// to the [removeSeparatedItem] separatorBuilder parameter.
+///
+/// An app that needs to insert or remove items in response to an event
+/// can refer to the [AnimatedListSeparated]'s state with a global key:
+///
+/// ```dart
+/// // (e.g. in a stateful widget)
+/// GlobalKey<AnimatedListSeparatedState> listKey = GlobalKey<AnimatedListSeparatedState>();
+///
+/// // ...
+///
+/// @override
+/// Widget build(BuildContext context) {
+///   return AnimatedListSeparated(
+///     key: listKey,
+///     itemBuilder: (BuildContext context, int index, Animation<double> animation) {
+///       return const Placeholder();
+///     },
+///     separatorBuilder: (BuildContext context, int index, Animation<double> animation) {
+///       return const Placeholder();
+///     },
+///   );
+/// }
+///
+/// // ...
+///
+/// void _updateList() {
+///   // adds "123" to the AnimatedListSeparated
+///   listKey.currentState!.insertItem(123);
+/// }
+/// ```
+///
+/// [AnimatedListSeparated] item input handlers can also refer to their [AnimatedListSeparatedState]
+/// with the static [AnimatedListSeparated.of] method.
+class AnimatedListSeparatedState extends _AnimatedScrollViewState<AnimatedListSeparated> {
```

Comment:
@goderbauer / @Piinks 
We probably shouldn't do it in this PR but the fact that we couldn't implement AnimatedListSeparatedState outside of the framework because _AnimatedScrollViewState is private probably means we have some magic here and should consider making _AnimatedScrollViewState public.